src/community/wps-download/src/main/java/org/geoserver/wps/gs/download/GridGeometryProvider.java

```diff
@@ -0,0 +1,395 @@
+
+    /** 
+     * Compute the requested GridGeometry taking into account resolution dimensions descriptor (if any),
+     * specified filter and ROI    
+     */
+    public GridGeometry2D getGridGeometry()
+            throws TransformException, IOException, FactoryException {
+        if (!StructuredGridCoverage2DReader.class.isAssignableFrom(reader.getClass())) {
+
+            //
+            // CASE A: simple readers: return the native resolution gridGeometry
+            //
+            if (LOGGER.isLoggable(Level.FINE)) {
+                LOGGER.fine("The underlying reader is not structured; returning native resolution");
+            }
+            return getNativeResolutionGridGeometry();
+        } else {
+            //
+            // CASE B: StructuredGridCoverage2DReader
+            // 
+            StructuredGridCoverage2DReader structuredReader = (StructuredGridCoverage2DReader) reader;
+            String coverageName = reader.getGridCoverageNames()[0];
+
+            Map<String, DimensionDescriptor> descriptors = structuredReader
+                    .getDimensionDescriptors(coverageName).stream()
+                    .collect(Collectors.toMap(dd -> dd.getName(), dd -> dd));
+
+            ResolutionProvider provider = new ResolutionProvider(descriptors);
+
+            //
+            // Do we have any resolution descriptor available?
+            // if not, go to standard computation.
+            //
+            if (!provider.canCompute()) {
+                if (LOGGER.isLoggable(Level.FINE)) {
+                    LOGGER.fine("The underlying reader is structured but no resolution descriptors are available"
+                            + ". Returning native resolution");
+                }
+                return getNativeResolutionGridGeometry();
+            }
+
+            GranuleSource granules = structuredReader.getGranules(coverageName, true);
+
+            // Setup a query on top of ROI and input filter (if any)
+            Query query = initQuery(granules);
+            SimpleFeatureCollection features = granules.getGranules(query);
+
+            if (features == null || features.isEmpty()) {
+                if (LOGGER.isLoggable(Level.FINE)) {
+                    LOGGER.fine("No features available for the specified query. Returning native resolution");
+                }
+                return getNativeResolutionGridGeometry();
+            }
+            // Initialize resolution with infinite numbers
+            double[] resolution = new double[] {Double.POSITIVE_INFINITY, Double.POSITIVE_INFINITY};
+            ReferencedEnvelope envelope = provider.getBestResolution(features, resolution);
+            AffineTransform at = new AffineTransform(resolution[0], 0, 0, -resolution[1], envelope.getMinX(), envelope.getMaxY());
+            MathTransform tx = ProjectiveTransform.create(at);
+            
+            return new GridGeometry2D(PixelInCell.CELL_CORNER, tx, envelope, GeoTools.getDefaultHints());
```

Comment:
What would happen if there are no granules captured by the query (and the resolutions remain at their infinite value)? I guess at some point an error should be returned of course, I'm just worried of whether the error would be understandable, or would look like a random internal error.

Reply:
Not sure to understand this comment.
Isn't the 
            if (features == null || features.isEmpty()) 
                return getNativeResolutionGridGeometry();
check above, already dealing with that?